crates/runtime/tests/acceleration/on_conflict.rs

```diff
@@ -0,0 +1,227 @@
+use std::time::Duration;
+
+use app::AppBuilder;
+use datafusion::assert_batches_eq;
+use futures::StreamExt;
+use runtime::{datafusion::query::Protocol, Runtime};
+use spicepod::component::{
+    dataset::{acceleration::Acceleration, Dataset},
+    params::Params,
+};
+
+use crate::{init_tracing, wait_until_true};
+
+#[cfg(feature = "postgres")]
+#[allow(clippy::too_many_lines)]
+#[tokio::test]
+async fn test_acceleration_on_conflict() -> Result<(), anyhow::Error> {
+    use arrow::array::RecordBatch;
+    use spicepod::component::dataset::acceleration::{OnConflictBehavior, RefreshMode};
+    use std::collections::HashMap;
+
+    use crate::get_test_datafusion;
+    use crate::postgres::common;
+
+    let _tracing = init_tracing(Some("integration=debug,info"));
+    let port: usize = 20963;
+    let running_container = common::start_postgres_docker_container(port).await?;
+
+    let pool = common::get_postgres_connection_pool(port).await?;
+    let db_conn = pool
+        .connect_direct()
+        .await
+        .expect("connection can be established");
+
+    db_conn
+        .conn
+        .execute(
+            "
+CREATE TABLE event_logs (
+    event_id SERIAL PRIMARY KEY,
+    event_name VARCHAR(100),
+    event_timestamp TIMESTAMP
+);",
+            &[],
+        )
+        .await
+        .expect("table is created");
+
+    db_conn
+        .conn
+        .execute(
+            "
+INSERT INTO event_logs (event_name, event_timestamp) VALUES
+('User Registration', '2023-05-16 10:00:00'),
+('Password Change', '2023-05-16 14:30:00'),
+('User Login', '2023-05-17 08:45:00'),
+('User Logout', '2023-05-17 18:00:00'),
+('File Download', '2023-05-17 13:20:00');",
+            &[],
+        )
+        .await
+        .expect("inserted data");
+
+    let params = Params::from_string_map(
+        vec![
+            ("pg_host".to_string(), "localhost".to_string()),
+            ("pg_port".to_string(), port.to_string()),
+            ("pg_user".to_string(), "postgres".to_string()),
+            ("pg_pass".to_string(), common::PG_PASSWORD.to_string()),
+            ("pg_sslmode".to_string(), "disable".to_string()),
+        ]
+        .into_iter()
+        .collect(),
+    );
+
+    // Create dataset with on_conflict:upsert acceleration setting
+    let mut on_conflict_upsert = Dataset::new("postgres:event_logs", "event_logs_upsert");
+
+    let mut upsert_hashmap = HashMap::new();
+    upsert_hashmap
+        .insert("event_id".to_string(), OnConflictBehavior::Upsert)
+        .unwrap_or_default();
+
+    on_conflict_upsert.params = Some(params.clone());
```

Comment:
Might be good to move dataset creation to helper function. There is duplicate logic between upsert and drop